iot_config/src/admin_service.rs

```diff
@@ -82,29 +101,51 @@ impl iot_config::Admin for AdminService {
 
         admin::insert_key(request.pubkey.clone().into(), key_type, &self.pool)
             .and_then(|_| async move {
-                self.auth_cache.insert_key(key_type, pubkey).await;
-                Ok(())
+                if self.auth_updater.send_if_modified(|cache| {
+                    if let std::collections::hash_map::Entry::Vacant(key) = cache.entry(pubkey) {
+                        key.insert(key_type);
+                        true
+                    } else {
+                        false
+                    }
+                }) {
+                    Ok(())
+                } else {
+                    Err(anyhow!("key already registered"))
+                }
             })
-            .map_err(|_| {
+            .map_err(|err| {
                 let pubkey: PublicKeyBinary = request.pubkey.into();
                 tracing::error!(pubkey = pubkey.to_string(), "pubkey add failed");
-                Status::internal(format!("error saving requested key: {pubkey}"))
+                Status::internal(format!("error saving requested key: {pubkey}, {err:?}"))
             })
             .await?;
 
-        Ok(Response::new(AdminKeyResV1 {}))
+        let timestamp = Utc::now().encode_timestamp();
+        let signer = self.signing_key.public_key().into();
+        let mut resp = AdminKeyResV1 {
+            timestamp,
+            signer,
+            signature: vec![],
+        };
+        resp.signature = self.sign_response(&resp.encode_to_vec())?;
+
+        Ok(Response::new(resp))
     }
 
     async fn remove_key(&self, request: Request<AdminRemoveKeyReqV1>) -> GrpcResult<AdminKeyResV1> {
         let request = request.into_inner();
 
-        self.verify_request_signature(&request).await?;
+        let signer = self.verify_public_key(&request.signer)?;
```

Comment:
why verify the public key, if we are subsequently verifying the signature ?  the sig verification will fail if the pub key is invalid anyway

Reply:
`verify_public_key` is how we turn the proto’s bytes field into a helium_crypto `PublicKey`, so we have to do some version of that verification in order to perform the signature verification

Reply:
k, i checked the actual verify_public_key function, its simply doing a try_from. I had originally thought there was more going on with this function